subscription.go

```diff
@@ -0,0 +1,219 @@
+	LifetimeCount              uint32
+	MaxKeepAliveCount          uint32
+	MaxNotificationsPerPublish uint32
+	Priority                   uint8
+	Notifs                     chan *PublishNotificationData
+}
+
+func NewMonitoredItemCreateRequestWithDefaults(nodeID *ua.NodeID, attributeID ua.AttributeID, clientHandle uint32) *ua.MonitoredItemCreateRequest {
+	if attributeID == 0 {
+		attributeID = ua.AttributeIDValue
+	}
+	return &ua.MonitoredItemCreateRequest{
+		ItemToMonitor: &ua.ReadValueID{
+			NodeID:       nodeID,
+			AttributeID:  attributeID,
+			DataEncoding: &ua.QualifiedName{},
+		},
+		MonitoringMode: ua.MonitoringModeReporting,
+		RequestedParameters: &ua.MonitoringParameters{
+			ClientHandle:     clientHandle,
+			DiscardOldest:    true,
+			Filter:           nil,
+			QueueSize:        10,
+			SamplingInterval: 0.0,
+		},
+	}
+}
+
+type PublishNotificationData struct {
+	SubscriptionID uint32
+	Error          error
+	Value          interface{}
+}
+
+// Cancel() deletes the Subscription from Server and makes the Client forget it so that publishing
+// loops cannot deliver notifications to it anymore
+func (s *Subscription) Cancel() error {
+	delete(s.c.subscriptions, s.SubscriptionID)
+
+	req := &ua.DeleteSubscriptionsRequest{
+		SubscriptionIDs: []uint32{s.SubscriptionID},
+	}
+	var res *ua.DeleteSubscriptionsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	if err != nil {
+		return err
+	}
+	if res.ResponseHeader.ServiceResult != ua.StatusOK {
+		return res.ResponseHeader.ServiceResult
+	}
+
+	return nil
+}
+
+func (s *Subscription) Monitor(ts ua.TimestampsToReturn, items ...*ua.MonitoredItemCreateRequest) (*ua.CreateMonitoredItemsResponse, error) {
+	// Part 4, 5.12.2.2 CreateMonitoredItems Service Parameters
+	req := &ua.CreateMonitoredItemsRequest{
+		SubscriptionID:     s.SubscriptionID,
+		TimestampsToReturn: ts,
+		ItemsToCreate:      items,
+	}
+
+	var res *ua.CreateMonitoredItemsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func (s *Subscription) Unmonitor(monitoredItemIDs ...uint32) (*ua.DeleteMonitoredItemsResponse, error) {
+	req := &ua.DeleteMonitoredItemsRequest{
+		MonitoredItemIDs: monitoredItemIDs,
+		SubscriptionID:   s.SubscriptionID,
+	}
+	var res *ua.DeleteMonitoredItemsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func publishOnceForSubscription(s *Subscription, acks []*ua.SubscriptionAcknowledgement) (*ua.PublishResponse, error) {
+	if acks == nil {
+		acks = []*ua.SubscriptionAcknowledgement{}
+	}
+	req := &ua.PublishRequest{
+		SubscriptionAcknowledgements: acks,
+	}
+
+	var res *ua.PublishResponse
+	err := s.c.sendWithTimeout(req, s.getPublishTimeout(), func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func (s *Subscription) getPublishTimeout() time.Duration {
```

Comment:
Can we rename this to just `publishTimeout()`?

Reply:
yep